DB-10882 Support updatable cursors

hatyo opened the pull request.

hatyo commented on splice_machine/src/main/java/com/splicemachine/derby/stream/function/ScrollInsensitiveFunction.java:
```diff
@@ -41,6 +41,9 @@ public ExecRow call(ExecRow execRow) throws Exception {
             this.operationContext.recordRead();
             op.setCurrentRow(execRow);
             this.operationContext.recordProduced();
+            if(!op.isForUpdate()) {
```
@dgomezferro this prevents serializing rowId. Should we added to other functions that could be pushed down to Spark executors and used by the `ResultStreamer`?

dgomezferro replied:
I think this is the only case, at least for now

hatyo replied:
ok

dgomezferro commented on db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/DeleteNode.java:
```diff
@@ -325,19 +325,11 @@ else if (cursorTargetTableName != null)
                                 COLUMNNAME,
-                                columnReference,
+                                rowIdColumn,
                                 getContextManager());
```
Why is this change needed? @ascend1 @msirek can you comment on whether this could cause issues?

hatyo replied:
If we don't expose the `CURRENT_ROW_LOCATION_NODE` in the RCL for update/delete nodes we will not generate the corresponding `getRowLocation` in the projection expression, which means we will not be able to retrieve the rowId from the projection node.
For more information please have a look at `com.splicemachine.db.impl.sql.compile.ResultColumnList#generateEvaluatedRow`

dgomezferro replied:
That makes sense, thanks!

dgomezferro commented on splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceIndexWatcher.java:
```diff
@@ -31,7 +31,7 @@
     protected String tableSchemaName;
     protected String createString;
-    protected String indexName;
+    public final String indexName;
```
Nitpick, a getter for this would be better

hatyo replied:
fixed it.

dgomezferro reviewed (changes requested):
Great job @hatyo , that was so fast! I only have some requests:
1. Can you add some more tests covering triggers (update/insert/delete) and concurrent updates? I've tested update triggers and they seem to work well, I'm interested in preventing regressions. Concurrent updates also seem to work well, if a row is deleted concurrently a WW conflict is raised for the update. After that the same cursor position can't be updated again, not sure if that's expected.
2. I think this is broken if we force Spark execution on the connection string : connect '...;useSpark=true' I get `ERROR 42X30: Cursor 'SQL_CURLH000C1' not found. Verify that autocommit is OFF.`

hatyo commented:
Thanks @dgomezferro for the comments, I didn't know we could enforce execution on Spark in the JDBC connection string, I will look into the failure and address the other comments.

hatyo commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

spliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2167/

hatyo commented:
I found some bugs in the implementation and fixed them. I also added some tests for interaction with triggers and concurrent transactions updating PK and non-PK columns.

spliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2173/

hatyo commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

spliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2174/

hatyo commented:
> TEST FAILED
> [http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2174](http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2174/)

Due to DB-9786.

hatyo commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

spliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2178/

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

spliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2179/

hatyo commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

spliceci commented:
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2181/

carolp-503 reviewed (approved):
+1

hatyo commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

dgomezferro commented on db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/IndexValueRow.java:
```diff
@@ -54,7 +54,7 @@
 public class IndexValueRow implements ExecIndexRow, Serializable {
 
-	private ExecRow valueRow;
+	private ExecRow                                     valueRow;
```
Remove extra spacing

hatyo replied:
done.

dgomezferro commented on splice_machine/src/main/java/com/splicemachine/derby/stream/function/SetCurrentLocatedRowAndRowKeyFunction.java:
```diff
@@ -17,27 +17,31 @@
     public ExecRow call(ExecRow locatedRow) throws Exception {
         getOperation().setCurrentRow(locatedRow);
         getOperation().setCurrentRowLocation(new HBaseRowLocation(locatedRow.getKey()));
+        getOperation().setCurrentBaseRowLocation(locatorFunction.call(operationContext));
```
We are setting this unconditionally, not just for "FOR UPDATE" statements. Is it possible to restrict it?

hatyo replied:
You're right, I restrict it for for-update plans.

dgomezferro commented on splice_machine/src/main/java/com/splicemachine/derby/stream/function/BaseRowLocatorFunction.java:
```diff
@@ -0,0 +1,55 @@
+import com.splicemachine.derby.impl.sql.execute.operations.TableScanOperation;
+import com.splicemachine.derby.stream.iapi.OperationContext;
+
+public class BaseRowLocatorFunction <Op extends SpliceOperation> extends SpliceFunction<Op, OperationContext<Op>,ExecRow> {
```
I don't think this needs to be a SpliceFunction, the logic could even be a static function. I'm ok with it being a functional object, but no need to extend SpliceFunction, implement `hasNativeSparkImplementation` etc.

hatyo replied:
You're right, I wanted initially to make it a full-fledged SpliceFunction that I inject in the plan execution, but along the way I reduced it to a normal static function. I will refactor it accordingly.

dgomezferro reviewed (changes requested):
Nice work! I have some comments, but I like the overall approach. Have you tested Partial Updates using indices in Spark?

hatyo commented on platform_it/pom.xml:
```diff
@@ -1069,6 +1069,7 @@
                                 <argument>-Dderby.language.updateSystemProcs=false</argument>
                                 <!-- Setting the logStatementText option to true enables logging of all statements. -->
                                 <argument>-Dderby.infolog.append=true</argument>
+                                <argument>-Dderby.debug.true=DumpClassFile</argument>
```
done.

hatyo commented:
> Nice work! I have some comments, but I like the overall approach. Have you tested Partial Updates using indices in Spark?

Thanks @dgomezferro for the comments, I addressed all of them and added a test for index scan with partial update.

ipraznik-splice commented:
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

spliceci commented:
TEST SUCCEEDED +1

spliceci commented:
TEST SUCCEEDED +1

dgomezferro reviewed (approved):
Awesome @hatyo , thank you!
